File: eaplanner/eaplanner/entities/schedule.py
import seaborn as sns
from pathlib import Path
import pickle
import random
from collections import defaultdict
from dataclasses import dataclass, field

import matplotlib.patches as patches
import networkx as nx
from matplotlib import pyplot as plt
from matplotlib.axes import Axes

from eaplanner.entities.assignment import Assignment
from eaplanner.entities.resource import Resource
from eaplanner.entities.constraint import (
    BaseConstraint,
    RelationConstraint,
    ResourceConstraint,
)
import csv


@dataclass
class Schedule:
    assignments: list[Assignment] = field(default_factory=list)
    constraints: list[BaseConstraint] = field(default_factory=list)

    def add_assignment(self, assignment: Assignment):
        self.assignments.append(assignment)

    def add_constraint(self, constraint: BaseConstraint):
        self.constraints.append(constraint)

    def get_total_makespan(self):
        if not self.assignments:
            return 0

        start = min(assignment.start for assignment in self.assignments)
        end = max(assignment.end for assignment in self.assignments)

        return end - start

    def get_start(self):
        return min(assignment.start for assignment in self.assignments)

    def get_end(self):
        return max(assignment.end for assignment in self.assignments)

    def get_total_penalty(self):
        return sum(constraint.get_penalty() for constraint in self.constraints)

    def get_assignment_by_id(self, id: int):
        return next(
            assignment for assignment in self.assignments if assignment.id == id
        )

    def reset_assignment_ids(self):
        for i, assignment in enumerate(sorted(self.assignments, key=lambda a: a.id)):
            assignment.id = i

    def remove_assignments_without_constraints(self):
        assignment_ids = set()
        for constraint in self.constraints:
            assignment_ids.update([a.id for a in constraint.get_assignments()])

        self.assignments = [a for a in self.assignments if a.id in assignment_ids]

    def repair_constraints(
        self, shuffle: bool = True, pct: float = 1.0, max_loops: int = 2
    ):
        constraints = self.constraints
        for _ in range(max_loops):
            # randomly select constraints
            constraints = [c for c in self.constraints if random.uniform(0, 1) <= pct]

            if shuffle:
                constraints = sorted(constraints, key=lambda _: random.random())

            # list comprehension to run all attempts
            if not any([c.attempt_repair() for c in constraints]):
                break

    def calculate_possible_combinations(self, duration: int | None = None):
        if duration is None:
            duration = self.get_total_makespan()

        try:
            return (duration * (duration - 1) / 2) ** len(self.assignments)
        except OverflowError:
            return float("inf")

    def get_constraint_per_group(self):
        groups: dict[str, list[BaseConstraint]] = defaultdict(list)
        for constraint in self.constraints:
            groups[type(constraint).__name__].append(constraint)

        return groups

    def get_total_hours_per_day(self):
        hours_per_day: dict[int, float] = defaultdict(float)
        for assignment in self.assignments:
            for day in range(assignment.start, assignment.end):
                hours_per_day[day] += assignment.hours_per_day

        return hours_per_day

    def as_graph(self):
        import networkx as nx

        G = nx.DiGraph()
        for assignment in self.assignments:
            G.add_node(
                assignment.id,
                start=assignment.start,
                duration=assignment.duration,
                hours=assignment.hours,
            )
        for constraint in self.constraints:
            if isinstance(constraint, RelationConstraint):
                G.add_edge(
                    constraint.predecessor.id,
                    constraint.successor.id,
                    penalty=constraint.get_penalty(),
                    label=constraint.type.short_name,
                )

        return G

    def visualize_graph(self, ax: Axes | None = None, show_edge_labels: bool = True):
        G = self.as_graph()

        if ax is None:
            _, ax = plt.subplots()

        pos = nx.nx_pydot.pydot_layout(G, prog="dot")
        penalties = list(nx.get_edge_attributes(G, "penalty").values())  # type: ignore
        hours = list(nx.get_node_attributes(G, "hours").values())  # type: ignore
        nx.draw(  # type: ignore
            G,
            pos,
            with_labels=True,
            edge_color=penalties,
            edge_cmap=plt.cm.RdYlGn_r,  # type: ignore
            node_color=hours,
            cmap=plt.cm.Reds,  # type: ignore
            ax=ax,
        )
        if show_edge_labels:
            nx.draw_networkx_edge_labels(  # type: ignore
                G,
                pos,
                edge_labels=nx.get_edge_attributes(G, "label"),  # type: ignore
                ax=ax,
            )

        return ax

    def visualize_gantt(
        self,
        ax: Axes | None = None,
        show_constraints: bool = True,
        show_legend: bool = False,
    ):
        if ax is None:
            _, ax = plt.subplots()

        ax.set_xlim(self.get_start() - 0.1, self.get_end() + 0.1)
        ax.set_ylim(0, max(assignment.id for assignment in self.assignments) + 1)

        for assignment in self.assignments:
            ax.add_patch(
                patches.Rectangle(
                    (assignment.start, assignment.id),
                    assignment.duration,
                    1,
                    linewidth=1,
                    zorder=1,
                )
            )

        if show_constraints:
            resource_constraints = [
                c for c in self.constraints if isinstance(c, ResourceConstraint)
            ]
            resource_colors = sns.color_palette(
                "colorblind", n_colors=len(resource_constraints)
            )
            for i, constraint in enumerate(resource_constraints):
                constraint.color = resource_colors[i]  # type: ignore
                # add to legend
                ax.add_patch(
                    patches.Rectangle(
                        (0, 0),
                        0,
                        0,
                        linewidth=1,
                        zorder=1,
                        label=constraint.resource.name,
                        color=constraint.color,
                    )
                )

            for constraint in self.constraints:
                ax = constraint.visualize_gantt(ax)

        # legend
        if show_legend:
            ax.legend()

        ax.set_xlabel("Day")
        ax.set_ylabel("Activity")
        ax.set_title("Gantt Chart")

        return ax

    def visualize_required_resources(self, ax: Axes | None = None, legend: bool = True):
        if ax is None:
            _, ax = plt.subplots()

        resources = [
            constraint
            for constraint in self.constraints
            if isinstance(constraint, ResourceConstraint)
        ]

        for resource in resources:
            daily_required_capacity = resource.get_daily_required_capacity()
            min_day = min(daily_required_capacity.keys())
            max_day = max(daily_required_capacity.keys())
            days = range(min_day, max_day + 1)
            capacity = [daily_required_capacity[day] for day in days]
            ax.plot(days, capacity, label=resource.resource.name)

        if legend:
            ax.legend()
        ax.set_xlabel("Day")
        ax.set_ylabel("Capacity")
        ax.set_title("Required capacity per resource")

        return ax

    def visualize_hours_per_day(self, ax: Axes | None = None):
        if ax is None:
            _, ax = plt.subplots()

        hours_per_day = self.get_total_hours_per_day()
        min_day = min(hours_per_day.keys())
        max_day = max(hours_per_day.keys())
        days = range(min_day, max_day + 1)
        hours = [hours_per_day[day] for day in days]
        ax.plot(days, hours)

        ax.set_xlabel("Day")
        ax.set_ylabel("Hours")
        ax.set_title("Hours per day")

        return ax

    def save(self, filename: Path):
        filename.parent.mkdir(parents=True, exist_ok=True)
        filename.write_bytes(pickle.dumps(self))

    def save_csv(self, filepath: Path):
        filepath.mkdir(parents=True, exist_ok=True)

        # write assignments
        with open(filepath / "activities.csv", "w+") as f:
            f.write("id,hours,start,duration\n")
            for assignment in self.assignments:
                f.write(assignment.to_csv_row() + "\n")

        # write relation constraints
        relation_constraints = [
            constraint
            for constraint in self.constraints
            if isinstance(constraint, RelationConstraint)
        ]
        with open(filepath / "sequence_constraints.csv", "w+") as f:
            f.write("predecessor_id,successor_id,type\n")
            for constraint in relation_constraints:
                f.write(constraint.to_csv_row() + "\n")

        # write resource constraints
        resource_constraints = [
            constraint
            for constraint in self.constraints
            if isinstance(constraint, ResourceConstraint)
        ]
        with open(filepath / "resource_constraints.csv", "w+") as f:
            f.write("resource_name,assignment_ids\n")
            for constraint in resource_constraints:
                f.write(constraint.to_csv_row() + "\n")

        # write resources
        resources = [c.resource for c in resource_constraints]
        with open(filepath / "resources.csv", "w+") as f:
            f.write("name,total_capacity\n")
            for resource in resources:
                f.write(resource.to_csv_row() + "\n")

    @staticmethod
    def load(filename: Path):
        if not filename.exists():
            raise FileNotFoundError(f"File {filename} does not exist")

        if filename.suffix == ".pkl":
            return Schedule.load_pickle(filename)

        return Schedule.load_csv(filename)

    @staticmethod
    def load_pickle(filename: Path):
        loaded = pickle.loads(filename.read_bytes())

        if not isinstance(loaded, Schedule):
            raise ValueError("File does not contain a Schedule object")

        return loaded

    @staticmethod
    def load_csv(filename: Path):
        assignments: list[Assignment] = []
        resources: list[Resource] = []
        constraints: list[BaseConstraint] = []

        with open(filename / "activities.csv", "r") as f:
            reader = csv.DictReader(f)
            for row in reader:
                assignments.append(Assignment.from_csv_row(row))

        with open(filename / "sequence_constraints.csv", "r") as f:
            reader = csv.DictReader(f)
            for row in reader:
                constraints.append(RelationConstraint.from_csv_row(row, assignments))

        with open(filename / "resources.csv", "r") as f:
            reader = csv.DictReader(f)
            for row in reader:
                resources.append(Resource.from_csv_row(row))

        with open(filename / "resource_constraints.csv", "r") as f:
            reader = csv.DictReader(f)
            for row in reader:
                constraints.append(
                    ResourceConstraint.from_csv_row(row, assignments, resources)
                )

        return Schedule(assignments=assignments, constraints=constraints)

    def __len__(self):
        return len(self.assignments)

File: eaplanner-rs/src/entities/schedule.rs
use rand::seq::SliceRandom;
use rand::Rng;
use std::path::Path;
use std::rc::Rc;

use super::assignment::Assignment;
use super::constraint::{BaseConstraint, RelationConstraint, ResourceConstraint};
use super::resource::Resource;
// use rand::prelude::*;

pub struct Schedule {
    pub assignments: Vec<Rc<Assignment>>,
    pub constraints: Vec<Box<dyn BaseConstraint>>,
}

impl Schedule {
    pub fn new() -> Self {
        Self {
            assignments: Vec::new(),
            constraints: Vec::new(),
        }
    }

    pub fn add_assignments(&mut self, assignments: Vec<Rc<Assignment>>) {
        for assignment in assignments {
            self.add_assignment(assignment);
        }
    }

    pub fn add_assignment(&mut self, assignment: Rc<Assignment>) {
        self.assignments.push(assignment);
    }

    pub fn add_constraints(&mut self, constraints: Vec<Box<dyn BaseConstraint>>) {
        for constraint in constraints {
            self.add_constraint(constraint);
        }
    }

    pub fn add_constraint(&mut self, constraint: Box<dyn BaseConstraint>) {
        self.constraints.push(constraint);
    }

    pub fn get_total_makespan(&self) -> i32 {
        self.get_end() - self.get_start()
    }

    pub fn get_start(&self) -> i32 {
        self.assignments
            .iter()
            .map(|a| a.start())
            .min()
            .unwrap_or_default()
    }

    pub fn get_end(&self) -> i32 {
        self.assignments
            .iter()
            .map(|a| a.end())
            .max()
            .unwrap_or_default()
    }

    pub fn get_total_penalty(&self) -> i32 {
        self.constraints
            .iter()
            .map(|c| c.get_penalty())
            .sum::<i32>()
    }

    pub fn repair_constraints(&self, pct: f32, max_loops: i32, shuffle: bool) {
        let mut rng = rand::thread_rng();
        let mut corrected = false;

        let mut order: Vec<usize> = (0..self.constraints.len()).collect();

        for _ in 0..max_loops {
            if shuffle {
                order.shuffle(&mut rng);
            }

            for i in order.iter() {
                if rng.gen::<f32>() < pct {
                    // false | false = false
                    corrected |= self.constraints[*i].attempt_repair();
                }
            }

            if !corrected {
                break;
            }
        }
    }

    pub fn get_constraints_per_type(&self) -> (Vec<&ResourceConstraint>, Vec<&RelationConstraint>) {
        let mut resource_constraints = Vec::new();
        let mut relation_constraints = Vec::new();

        for constraint in self.constraints.iter() {
            if let Some(resource_constraint) =
                constraint.as_any().downcast_ref::<ResourceConstraint>()
            {
                resource_constraints.push(resource_constraint);
            } else if let Some(relation_constraint) =
                constraint.as_any().downcast_ref::<RelationConstraint>()
            {
                relation_constraints.push(relation_constraint);
            }
        }

        (resource_constraints, relation_constraints)
    }

    pub fn to_csv(&self, save_folder: &Path) {
        // write assignments
        let assignments_filepath = save_folder.join("activities.csv");
        let assignments_file =
            std::fs::File::create(assignments_filepath).expect("Failed to create file");
        let mut assignments_writer = csv::Writer::from_writer(assignments_file);
        assignments_writer
            .write_record(&["id", "hours", "start", "duration"])
            .expect("Failed to write header");

        for assignment in self.assignments.iter() {
            assignments_writer
                .write_record(&[
                    assignment.id().to_string(),
                    assignment.hours().to_string(),
                    assignment.start().to_string(),
                    assignment.duration().to_string(),
                ])
                .expect("Failed to write data");
        }

        // get the different constraints
        let (resource_constraints, relation_constraints) = self.get_constraints_per_type();

        // write resources
        let resources_filepath = save_folder.join("resources.csv");
        let resources_file =
            std::fs::File::create(resources_filepath).expect("Failed to create file");
        let mut resources_writer = csv::Writer::from_writer(resources_file);
        resources_writer
            .write_record(&["name", "total_capacity"])
            .expect("Failed to write header");

        let resources = resource_constraints
            .iter()
            .map(|c| c.resource.clone())
            .collect::<Vec<Rc<Resource>>>();
        let mut unique_resources = Vec::new();
        for resource in resources.iter() {
            if !unique_resources.contains(resource) {
                unique_resources.push(resource.clone());
            }
        }
        for resource in unique_resources.iter() {
            resources_writer
                .write_record(&[resource.name().to_string(), resource.capacity().to_string()])
                .expect("Failed to write data");
        }

        // write relation constraints
        let relations_filepath = save_folder.join("sequence_constraints.csv");
        let relations_file =
            std::fs::File::create(relations_filepath).expect("Failed to create file");
        let mut relations_writer = csv::Writer::from_writer(relations_file);
        relations_writer
            .write_record(&["predecessor_id", "successor_id", "type"])
            .expect("Failed to write header");

        for constraint in relation_constraints.iter() {
            relations_writer
                .write_record(&[
                    constraint.predecessor.id().to_string(),
                    constraint.successor.id().to_string(),
                    (constraint.r#type as u8).to_string(),
                ])
                .expect("Failed to write data");
        }

        // write resource constraints
        let resource_constraints_filepath = save_folder.join("resource_constraints.csv");
        let resource_constraints_file =
            std::fs::File::create(resource_constraints_filepath).expect("Failed to create file");
        let mut resource_constraints_writer = csv::Writer::from_writer(resource_constraints_file);
        resource_constraints_writer
            .write_record(&["resource_name", "assignment_ids"])
            .expect("Failed to write header");

        for constraint in resource_constraints.iter() {
            let assignments = constraint
                .assignments
                .iter()
                .map(|f| f.id().to_string())
                .collect::<Vec<String>>()
                .join(";");
            resource_constraints_writer
                .write_record(&[constraint.resource.name(), assignments])
                .expect("Failed to write data");
        }
    }
}
